libsolidity/analysis/TypeChecker.cpp

```diff
@@ -116,7 +116,7 @@ void TypeChecker::checkDoubleStorageAssignment(Assignment const& _assignment)
 		TupleType const& lhsType = dynamic_cast<TupleType const&>(*type(_lhs));
 		TupleExpression const* lhsResolved = dynamic_cast<TupleExpression const*>(resolveOuterUnaryTuples(&_lhs));
 
-		if (lhsType.components().size() != _rhs.components().size() || lhsResolved->components().size() != _rhs.components().size())
+		if (!lhsResolved || lhsType.components().size() != _rhs.components().size() || lhsResolved->components().size() != _rhs.components().size())
```

Comment:
I think this would make more sense:

```suggestion
		if (lhsType.components().size() != _rhs.components().size() || (lhsResolved && lhsResolved->components().size() != _rhs.components().size()))
```
Your version makes an implicit assumption that when `lhsResolved` is null, the code cannot be correct. While this assumption *is* true, that's only really because currently you cannot return tuple L-values from functions. We could easily avoid this hidden assumption.

Moreover, the second part of the condition seems superfluous to me. I think that non-null `lhsResolved` will always match the number of components in `lhsType`, so I'd actually change it to this:

```suggestion
		solAssert(!lhsResolved || lhsResolved->components().size() == lhsType.components().size());
		if (lhsType.components().size() != _rhs.components().size())
```

Reply:
Actually, let me explain it in more detail, because there are a few gotchas about our type system that might not be obvious and even knowing these details it took me a while to think through the possible cases here.

We're dealing with `TupleType` and `TupleExpression` here and, perhaps counterintuitively, one does not imply the other. Something can be of `TupleType` even if it's not a `TupleExpression`. For example a `FunctionCall` expression `f()` will get a tuple type when the function returns multiple values or zero values. And conversely, something can be a `TupleExpression` even if it's not of a `TupleType`. For example `()`, `(1)` and `(1, 1)` are all tuple expressions, but only the first and the last are of tuple type. `(1)` is a parenthesized rational number. Parenthesized expressions are the only special case - in that case the type comes from the content. So e.g. `(1)`, `((((1, 2, 3)))` and `((f())` have the same types as `1`, `(1, 2, 3)` and `f()`, respectively. Note that an empty tuple is of `TupleType` even though it has no content.

So now, `TypeChecker::checkDoubleStorageAssignment()` assumes that both sides of the assignment are of `TupleType` (I think it should really have explicit asserts against that). Otherwise it won't get called. This filters out some possibilities for LHS (e.g. cases like `(x) = (1)` that would otherwise fail the condition here). We're left with:
1. Not a single element tuple expression. The type comes directly from that expression and has the same number of components.
    ```solidity
    function f() public {
        (uint x, uint y) = (1, 1); // OK
    }
    ```
1. Single element tuple expression containing a non-single element tuple expression. The type comes directly from that nested expression and has the same number of components.
    ```solidity
    function f() public {
        ((x.push(), x.push())) = (1, 1); // OK
    }
    ```
1. Single element tuple expression containing something else of tuple type (e.g. a function returning a tuple). In that case `lhsResolved` is null.
    ```solidity
    function f() public returns (uint, uint) {
        (f()) = (1, 1); // Error: Expression has to be an lvalue.
    }
    ```

For that reason checking the number of components in `lhsResolved` is redundant. It either follows from the number of components in the type or is irrelevant.

Reply:
Right. Thanks for the explanation, I wasn't aware of such distinction ;)
I applied the suggestion and also your examples as test cases.